paasta_tools/setup_istio_mesh.py

```diff
@@ -0,0 +1,245 @@
+        for namespace, endpoint in smartstack_namespaces.items():
+            paasta_namespaces[endpoint["port"]] = namespace
+    return paasta_namespaces
+
+
+def sanitise_kubernetes_service_name(name: str) -> str:
+    name = sanitise_kubernetes_name(name)
+    return name.replace(".", "--")
+
+
+def get_existing_kubernetes_service_names(kube_client: KubeClient) -> Set[str]:
+    label_selector = "paasta.yelp.com/owner=istio"
+    service_objects = kube_client.core.list_namespaced_service(
+        "paasta", label_selector=label_selector
+    )
+    if not service_objects:
+        raise ErrorGettingServiceList("Error retrieving services list from k8s api")
+
+    return {item.metadata.name for item in service_objects.items}
+
+
+def setup_unified_service(
+    kube_client: KubeClient, port_list: AbstractSet
+) -> k8s.V1Service:
+
+    # Add port 1337 for envoy unified listener.
+    # Clients can connect to this listenner and set x-yelp-svc header for routing
+    ports = [
+        k8s.V1ServicePort(
+            name=f"p{UNIFIED_SVC_PORT}",
+            port=UNIFIED_SVC_PORT,
+            protocol="TCP",
+            target_port=PAASTA_SVC_PORT,
+            app_protocol="http",
+        )
+    ]
+
+    # Add smartstack ports for routing, Clients can connect to this
+    # Directly without need of setting x-yelp-svc header
+    for port in port_list:
+        port_spec = k8s.V1ServicePort(
+            name=f"p{port}",
+            port=port,
+            protocol="TCP",
+            target_port=PAASTA_SVC_PORT,
+            app_protocol="http",
+        )
+        ports.append(port_spec)
+
+    service_labels: KubeSvcLabels = {
+        "paasta.yelp.com/owner": "istio",
+        "paasta.yelp.com/unified_service": "true",
+    }
+
+    service_meta = k8s.V1ObjectMeta(name=UNIFIED_K8S_SVC_NAME, labels=service_labels)
+    service_spec = k8s.V1ServiceSpec(ports=ports)
+    service_object = k8s.V1APIService(metadata=service_meta, spec=service_spec)
+    return kube_client.core.create_namespaced_service(PAASTA_NAMESPACE, service_object)
+
+
+def setup_paasta_namespace_service(
+    kube_client: KubeClient,
+    paasta_namespaces: Mapping,
+    existing_kube_services_names: Set[str] = set(),
+    rate_limit: int = 0,
+) -> bool:
+    api_updates = 0
+    status = True
+
+    for namespace in paasta_namespaces.values():
+        service = sanitise_kubernetes_service_name(namespace)
+        if rate_limit > 0 and api_updates >= rate_limit:
+            log.info(
+                f"Not doing any further updates as we reached the limit ({api_updates})"
+            )
+            break
+
+        if service in existing_kube_services_names:
+            log.info(f"Service {service} alredy exists, skipping")
+            continue
+
+        log.info(f"Creating {service} because it does not exist yet.")
+
+        service_labels: KubeSvcLabels = {
+            "paasta.yelp.com/owner": "istio",
+            "paasta.yelp.com/unified_service": "false",
```

Comment:
i don't see where we use this label, isn't matching by service name good enough? we can assume that all services created under paasta namespace are part of our mesh setup, nothing else should be creating services in here.